localize-currency-amount: Utility for locale-aware currency formatting

nbloomf opened the pull request:
Formatting monetary amounts depends on the currency and the user's locale. This package adds a function for dealing with this.

#### Changes proposed in this Pull Request

* Adds `localizeMonetaryAmount()`. This improves on other methods by working with integers only and using non breaking spaces everywhere.

#### Testing instructions

`npm run test-packages localize-monetary-amount`

sirbrillig commented on packages/wp-checkout/src/lib/format-monetary-amount-for-locale.js:
```diff
@@ -0,0 +1,781 @@
+/**
+ * External dependencies
+ */
+import { isString } from 'lodash';
```
I think we can remove this dependency. I'd like to figure out what problems we could encounter with `typeof`.

nbloomf replied:
Removed.

sirbrillig commented on packages/wp-checkout/src/lib/format-monetary-amount-for-locale.js:
```diff
@@ -0,0 +1,781 @@
+import { isString } from 'lodash';
+
+/**
+ * Format a monetary amount according to the custom of a given locale.
```
This comment block is beautiful. ❤️

sirbrillig commented on packages/wp-checkout/src/lib/format-monetary-amount-for-locale.js:
```diff
@@ -0,0 +1,781 @@
+ * @param {string} localeCode ISO 3166-1 alpha-2 region code
+ * @returns {{decimalSeparator:string, groupSeparator:string}} Decimal and group separators
+ */
+function separatorsForLocale( localeCode ) {
```
I have no problem with having a custom one, but we could maybe make use of the `Intl.NumberFormat('ja-JP', { style: 'currency', currency: 'JPY' }).format(number)` function to format separators and currency symbols?

nbloomf replied:
I didn't know about `Intl.NumberFormat` -- thanks for the pointer.

Just musing about this... I'm torn because on one hand that would handle all of the configuration around separators and grouping conventions, but we'd also lose some flexibility (which is funny because that function takes like 12 options) around currency symbols. For instance, in Australia, should Australian dollars be formatted as `A$ 10` or `A$10` or `$10 AUD`? Does it matter? I don't know. But `toLocaleString` picks this for us.

Also the `NumberFormat` object uses floats, while we're trying to stick with integer arithmetic. I can think of a couple of ways to deal with that. One is to just use a float; divide by the minor unit right before passing the amount to `NumberFormat` and use options to round. This brings us back to the accuracy issue with floats, albeit not nearly as severe because we only divide once at the end by a small integer. It hurts my soul to settle for that when an exact solution is right there, but so it goes. :) Another option is to extract the major unit portion as an integer, format it (as a float), and manually replace the decimal part of the string, but that smells terrible.

sirbrillig replied:
You make a good case. It's actually a little surprising to me that the built-in methods use floats.  It seems like a big oversight. I suppose the answer is probably that they didn't want to have to include data for major/minor unit conversion. Well, do what you think best! As long as the tests pass we can always change the underlying implementation later.

sirbrillig reviewed (commented):
This looks amazing. I think we can actually make this into its own package and then depend on it.

sirbrillig commented:
I also just saw this, but I don't know much about it: https://github.com/Automattic/wp-calypso/tree/master/packages/format-currency

Commit:
Add a package for localizing monetary amounts

This library was written for the in-progress composite-checkout package but has no external dependencies, so we're extracting it as its own package. Features: exact arithmetic and non-breaking spaces

sirbrillig commented:
I think this is great. My inclination is to wait until #36720 is merged and then make this PR into its own package (maybe `@automattic/format-amount-for-locale`?) and add it to the dependencies of `@automattic/composite-component`.

nbloomf commented:
That sounds good to me.

sirbrillig commented on packages/localize-monetary-amount/src/localize-monetary-amount.ts:
```diff
@@ -0,0 +1,980 @@
+): LocalizedMonetaryAmount {
+	// Validate arguments
+	if ( ! ( typeof amount === 'number' && Number.isInteger( amount ) ) ) {
+		throw 'amount parameter must be an integer.';
```
Generally [it's best](https://stackoverflow.com/questions/11502052/throwing-strings-instead-of-errors) to throw an `Error` object (or a custom error if you like); many places that catch errors expect this, and it gives helpful metadata (like a stack trace) for free.

sirbrillig commented on packages/localize-monetary-amount/src/localize-monetary-amount.ts:
```diff
@@ -0,0 +1,980 @@
+				.concat( '\u00A0'.concat( currencyCode ) );
+
+		default:
+			throw `Unrecognized currency format ${ schema }`;
```
Same comment about throwing an `Error`.

sirbrillig commented on packages/localize-monetary-amount/src/localize-monetary-amount.ts:
```diff
@@ -0,0 +1,980 @@
+	absoluteAmount: number
+): { integerPart: string[]; fractionalPart: string } {
+	if ( ! ( Number.isInteger( absoluteAmount ) && absoluteAmount >= 0 ) ) {
+		throw 'absoluteAmount parameter must be a nonnegative integer.';
```
Same throw comment.

sirbrillig commented on packages/localize-monetary-amount/src/localize-monetary-amount.ts:
```diff
@@ -0,0 +1,980 @@
+		case 'CAD': // Canadian dollar
+		case 'USD': // US dollar
+			if ( absoluteAmount < 0 || 100 <= absoluteAmount ) {
+				throw 'Minor units should be between 0 and 99 inclusive';
```
Same throw comment.

sirreal commented on packages/localize-monetary-amount/test/localize-monetary-amount.js:
```diff
@@ -0,0 +1,358 @@
+		} );
+		it( 'en-us/USD, 50', function() {
+			expect( localizeMonetaryAmount( 'en-us', 'USD', 50 ) ).toBe( '$0.50' );
+		} );
```
This doesn't need to be rewritten, but these very repetitive tests can be written concisely and expressively with [```test.each`table`(…)```](https://jestjs.io/docs/en/api#2-testeach-table-name-fn-timeout).

nbloomf replied:
TIL!

sirreal reviewed (commented):
This is some impressive work, well done 💯 

I did my best to provide a thorough review. I learned a lot in the process 🙂 

A few high level notes:
- We should copy `.eslintrc` from a sibling package. That's required for one of our package dependency lint rules.
- We need to get the build set up and I'd like to use the `tsc` compiler to output types.
- Let's remove types from jsdoc, they're redundant.
- We need to check browser compatibility. If we're targetting es5, we should use es5 or make sure that code is properly transpiled. I don't know whether TypeScript handles that.
- The structure and comments are impressive and make this pretty easy to follow even when it's completely new material. Well done 👍

Commit:
Incorporate feedback

nbloomf commented on packages/localize-monetary-amount/src/localize-monetary-amount.ts:
```diff
@@ -0,0 +1,980 @@
+/**
+ * Representation of the possible currency schemas.
+ */
+enum CurrencyFormat {
```
TIL about the difference between `enum` and `const enum`. :)

nbloomf commented on packages/localize-monetary-amount/src/localize-monetary-amount.ts:
```diff
@@ -0,0 +1,980 @@
+ * Localize a monetary amount using the given schema slug.
+ *
+ * @param {number} amountSign
+ *   1, 0, or -1, to distinguish positive and negative amounts
```
In this library we really do only want the "sign" and not to worry about negative zero since this package only deals with integers. I ended up defining a `const enum` type to represent the sign, so we can define away float weirdness. :)

sirreal replied:
Nice solution 👌

nbloomf commented on packages/localize-monetary-amount/src/localize-monetary-amount.ts:
```diff
@@ -0,0 +1,980 @@
+	schema: CurrencyFormat
+): LocalizedMonetaryAmount {
+	const symbol = localSymbolForCurrency( localeCode, currencyCode );
+	const sign = amountSign >= 0 ? '' : '-';
```
The latest version uses a specialized type to represent the sign, so we can say

```
const signSymbol = sign === Sign.IsNegative ? '-' : '';
```

nbloomf commented on packages/localize-monetary-amount/src/localize-monetary-amount.ts:
```diff
@@ -0,0 +1,980 @@
+	switch ( schema ) {
+		// $1,000.50
+		case CurrencyFormat.LocalSymbol_Amount:
+			if ( 0 === amountSign ) {
```
The latest commit uses the `Sign` enum type makes the intent here more clear and avoid negative zero.

sirreal commented on packages/localize-monetary-amount/src/localize-monetary-amount.ts:
```diff
@@ -0,0 +1,980 @@
+		// 10 000.00
+		case 'en-au': // English - Australia
+		case 'en-ca': // Canada
+			return { decimalSeparator: '.', groupSeparator: '\u00A0' };
```
With `"\u00A0"`, do you think it's worth extracting to a constant `NO_BREAK_SPACE` or `NBSP`?

nbloomf replied:
Good idea - this makes the intent more clear.

Commit:
Incorporate feedback

Commit:
Add tsconfig.json and update package.json

jblz commented:
I just did a cursory scan of this, so please forgive my lack of context.

What's the driver behind putting this in the client? Won't monetary data be coming from the API? What's preventing us from formatting it before we send it to the client (e.g. with our CLDR functionality)?

nbloomf commented:
Thanks for taking a look @jblz!

> What's preventing us from formatting it before we send it to the client?

Nothing, and WPCOM does this already. But this package is an offshoot of [composite-checkout](https://github.com/Automattic/wp-calypso/tree/master/packages/composite-checkout), which will be used on other services. That package requires its host client to provide its own formatted currency strings, and this package allows them to do so without changing their own API if needed.

sirreal commented:
Curious if this is still planned to move forward. There appears to be a lot of valuable work here!

sirbrillig commented:
I think it would be great if this work could be shared. Maybe we could merge and publish this package, regardless of our current use of it? Or it could be published outside of the calypso monorepo.

sirreal commented:
With #39786 merged, this will likely require a few build-related changes, but shouldn't be too much work.

Just getting this merged would be a great first step 😁

sirbrillig commented:
@nbloomf What do you think?

sirbrillig commented:
I still think this would be great to have. We have a bunch of places in calypso where we're forced to get a lot of extra data from the server to avoid client-side math and this would resolve that need.

sirbrillig commented:
I still like it
